### PR TITLE
Add color support to rgbgfx (again)!

### LINKED ISSUE #232: Rgbgfx doesn't handle color yet
`rgbgfx` is a bit incomplete. Although there's a command line switch to export a palette file from a given PNG, and the documentation mentions "sorting colors from lightest to darkest", in actuality `rgbgfx` doesn't take color into account at all. Instead, it currently [quantizes](https://github.com/rednex/rgbds/blob/1a5c423984ca6cce1cb0b6b2100836bf4aa56e84/src/gfx/makepng.c#L153-L157) colors to the first 4 colors in the palette if the PNG is indexed, and to a preset grayscale palette if truecolor or grayscale. The documentation and [a few comments](https://github.com/rednex/rgbds/blob/1a5c423984ca6cce1cb0b6b2100836bf4aa56e84/src/gfx/makepng.c#L106-L111) suggest that this was meant to be temporary behavior until the much more complicated intended logic had been implemented. It makes sense to implement, too - currently, the palette file output does nothing (except output the preset grayscale palette) for non-indexed images, and truecolor images are often quantized in a way that assigns the same palette index to multiple colors.

There are also other color-related problems that `rgbgfx` doesn't handle properly yet - for example, transparency.

In practice, this means a few features have to be implemented:
* Truecolor and grayscale PNGs need to be analyzed to build a palette from the pixel colors.
* Generated palettes need to be sorted by relative luminance.
* All types of PNGs need to be checked for palettes with too many colors.
* Transparent pixels need to correspond to a color in the palette, as opposed to adding a new one or being thrown away.

All of these features can be implemented in a way that's backwards-compatible with *at least* pokecrystal and [Telefang](https://github.com/telefang/telefang), and with what the documentation states, so implementing / fixing these things shouldn't cause any problems.

I've implemented the missing features on my end, so I'll have a pull request coming in shortly (after a bit of extra polish)!

### PR DESCRIPTION
**Pull request by obskyr**
Repost of #234, as discussed. Fixes #232, once again.

### CODE REVIEW

**Comment by obskyr** [comment]
I couldn't literally squash the second style fix commit, as there was a commit in between the two, but I integrated the changes manually into the former commit. I really do dislike faking history (I don't see the point), but I don't feel like arguing about it, hahah. Everything should be all good now.

**Comment by AntonioND** [comment]
You can do pretty much anything with your commits: reorder them, change the message, join them... Allow me to introduce you `git rebase interactive`. :P https://robots.thoughtbot.com/git-interactive-rebase-squash-amend-rewriting-history

Having a clean history is really useful:
- `git bisect` actually gives you a clean commit that introduced the bug, which reduces the time you need to find it.
- Similarly, `git blame` is more useful because you don't have a lot of useless changes all over the code, just the changes that you actually need.
- `git revert` works better because you don't have to worry so much about changes that happened between the commit that introduced a bug and the master branch, it is likely that there will be less commits that modified the affected code.

In short, for development it is fine if you have a messy history but, once the code goes to the upstream repository, having a clean history is worth the effort.

**Comment by obskyr** [comment]
@AntonioND Of course I know how an interactive rebase works – I had to `git rebase -i` about 5 times just for this PR! Squashing merges the commit with the *previous* one, which wouldn't have worked here, as there was a commit (the superfluous height commit) *between* the two style commits.

I can't say I agree with your views on git history – for me, preserving an accurate history, taking dramatically less time for busywork (after all, you want to spend your time *developing*, not juggling rebases and merge conflicts like I've just spent a few days doing), and not scaring away contributors trumps any advantage a "clean" history might give.

But hey, whatever the case, I did it. Bam.

**Comment by AntonioND** [comment]
You can reorder commits with `rebase interactive`:
```
pick A
pick B
pick C
```
Can be modified to:
```
pick A
squash C
pick B
```

It is easier from the developer's point of view, but not the maintainer's one.

I think that what you say is an ok(ish) strategy if you don't care too much about the history because you're prototyping or you're just working on your own and nobody else is going to touch that repository because you have a mental map of everything (until you forget about something, that's when the problems start). If not, if someone else has to fix something, having a messy history can turn a 30 minutes debug task into a 3 hour one.

In a repository like this, having a clean history is much more valuable than having a faster development pace because, well, most of the current work in this project is maintenance work, and there are other projects that depend on it so it's better to help the people that maintain them, it's easier for everyone at the expense of a bit more of effort from the developer.

I mean, if someone decided to create a PR with just one commit of 2000 lines I wouldn't even take a look at it until the author decided to split it in logical chunks. Or if there were 10 commits but all of the commits were reworks of the previous one. However, in that case, I could always squash them if the result wasn't too big.
